modules/assim.sequential/R/Analysis_sda.R

```diff
@@ -272,34 +271,123 @@ GEF<-function(setting,Forecast,Observed, H, extraArg, nitr=50000, nburnin=10000,
   #### from the interval matrix
   y.ind <- as.numeric(Y > interval[,1])
   y.censored <- as.numeric(ifelse(Y > interval[,1], Y, 0))
-
-  if(t == 1){ #TO DO need to make something that works to pick whether to compile or not
-    # Contants defined in the model
-    constants.tobit = list(N = ncol(X), YN = length(y.ind))
+  
+  #which type of observation do we have at this time point?
+  input.order <- lapply(input.vars, grep, x=names(obs.mean[[t]])) # not going to work if AbvGrnWood is given in two different ways like tree rings and refab
+  names(input.order) <- operators
+  data_available <- unlist(input.order)
+  
+  #browser()
+  
+  if(any(grep(names(data_available),pattern = 'direct'))){
+    which_direct <- data_available[grep(names(data_available),pattern = 'direct')]
+    X_direct_start <- which_direct[1]
+    X_direct_end <- which_direct[length(which_direct)]
+    direct_TRUE = TRUE
+  }else{
+    X_direct_start <- 0
+    X_direct_end <- 0
+    direct_TRUE = FALSE
+  }
+  
+  if(any(grep(names(data_available),pattern = 'ALR'))){
+   
+    # browser()
+    which_fcomp <- grep(names(data_available),pattern = 'ALR')
+    X_fcomp_start <- which_fcomp[1]
+    X_fcomp_end <- which_fcomp[length(which_fcomp)]
+    X_fcomp_model <- grep(colnames(X),pattern = 'AbvGrndWood')
     
+    if(FALSE){
+      X_fcomp_idx <- pft_order(X = X,Y = Y,var.names = var.names,input.vars = input.vars)
+      if(length(X_fcomp_idx)!=X_fcomp_idx[length(X_fcomp_idx)]){
+        X_fcomp_idx <- c(X_fcomp_idx,which(is.na(match(1:X_fcomp_idx[length(X_fcomp_idx)],X_fcomp_idx))))
+      }
+    }
+  
+    fcomp_TRUE = TRUE
+  }else{
+    X_fcomp_start <- 0
+    X_fcomp_end <- 0
+    X_fcomp_model <- 0
+    fcomp_TRUE = FALSE
+  }
+  
+  if(any(grep(names(data_available),pattern = 'pft2total'))){
```

Comment:
Is there a way to generalize these blocks so that we don't have to add new code like this every time we add a new observation operator?